### DISCUSSION TITLE
blobsaver v3.0 Beta

### DISCUSSION THREAD

**Discussion by airsquared** [airsquared/blobsaver#242] [General]
# The full release of v3.0 is out, please download it from [here](https://github.com/airsquared/blobsaver/releases).

-----
<s>
There's some major changes in this version, so please report any feedback or issues you encounter with this version here.

Note that: 

- This new version stores saved devices differently, so you will need to import your presets from older versions of blobsaver using the "File" menu if you would like to restore them.
- The version number says v2.6, but it's actually a 3.0 beta. This is to differentiate from when the full version is released.
- Because it's a beta, it will prompt you that a new update (v2.5.5) is available, even though that's an older version. To avoid the prompt, you can just click "ignore this version" the next time you see it.
- When the normal release of v3.0 is out, it will show a similar prompt to upgrade to v3.0 which you can download from the releases tab like normal.
- Java is no longer required starting in the v3.0 beta, so you can uninstall it if you like

Downloads:
macOS: [`blobsaver-2.6.dmg`](https://github.com/airsquared/blobsaver/raw/beta/binaries/blobsaver-2.6.dmg) (if it says the app cannot be opened, try [this](https://github.com/airsquared/blobsaver/wiki/Unable-to-open-on-macOS-Catalina))
Windows: [`blobsaver-2.6.exe`](https://github.com/airsquared/blobsaver/raw/beta/binaries/blobsaver-2.6.exe)
Linux: [`blobsaver_2.6-1_amd64.deb`](https://github.com/airsquared/blobsaver/raw/beta/binaries/blobsaver_2.6-1_amd64.deb), [`blobsaver-2.6-1.x86_64.rpm`](https://github.com/airsquared/blobsaver/raw/beta/binaries/blobsaver-2.6-1.x86_64.rpm), [`blobsaver-2.6.tar.gz`](https://github.com/airsquared/blobsaver/raw/beta/binaries/blobsaver-2.6.tar.gz)

## Changelog

- Read *both* the generator and APNonce from devices, so blobs can be saved on unjailbroken devices with a known generator/ApNonce pair (major thanks to @nyuszika7h for discovering this!)
- Now uses the system scheduler for saving blobs in the background instead of having blobsaver run constantly in the background, reducing system resource usage (launchd on macOS, task scheduler on Windows, systemd timers on Linux)
- Presets are now called saved devices, and support an unlimited number of devices
- Added feature to export/import saved devices to/from XML
- Allow optionally specifying a generator
- Fix getting stuck in recovery mode when trying to read APNonce
- Improved performance and error handling
- Added support for some devices that I missed previously
- Java is now included and does not require a separate installation (also updated to Java 16)
- UI updates
- Add basic & anonymous analytics that can be disabled from the "Help" menu
- A lot of other misc. fixes and enhancements + major code cleanup/reorganization

</s>

**Comment by EdgeXS** [comment]
Thanks for this. Just downloaded.

Issues:
- install: it didn't allow to run on the mac and i had to resign it using sudo codesign --force --deep --sign - /Applications/blobsaver.app
- when adding multiple devices. The force check is greyed out. (had to restart a few times before the force check wasn't greyed out)
- when background mode is enabled there is no menulet blobsaver icon however the activity monitor does the process running not sure if this is by design?
- when in background mode, if you reopen the app you don't get the blobsaver window.  This is even true when you select the window nothing happens. But if you close it a shadow window appears before quitting.  Reopening app allows you to see the window again.

Working
- background process seems to be working: when 1 minute frequency is created it goes and creates blob for devices for multiple devices.

**Reply by airsquared** [reply]
1. Did the usual procedure of right click > Open not work? I'm not able to reproduce this.
2. Just fixed this, try downloading the latest build again
3. This is intentional, with the new update blobsaver does not run all the time in the background and is only running when it is actually saving blobs
4. I'm not able to reproduce this one either, do you have a video/screenshot?

**Reply by EdgeXS** [reply]
1. I tried on 10.14.6 it didn't work but did work on 10.15.
2. Didn't work initially but can't reproduce it.  Might be an upgrade issue.
3. Yes, I can confirm that there is no process or menulet running but the background process does work. 
4. Seems like background mode has nothing to do with it. I isolated to the app itself.  When you open the app it doesn't show the window. If you select the window you get an error sound.  When you quit the app you see the window briefly.  Repeating this process in 10.14.6 it eventually shows up.  Video: https://www.dropbox.com/s/qerqcjpl0lou5un/blobsaver-reopening.mp4?dl=0

**Reply by airsquared** [reply]
Does this bug with the window happen on 10.15 too, or just 10.14.6?

**Reply by joostiphone** [reply]
I had exactly the same issue not able to open it (in Dutch MacOS (11.2.3) says it's damaged and needs to be removed). So I used the command @EdgeXS mentioned and now it opens using control+open.

**Reply by airsquared** [reply]
It seems that I can't do anything about it saying the app being damaged, unless I pay for an Apple Developer account to sign the app. I just added a link to the wiki page with a solution.

**Reply by CoocooFroggy** [reply]
> It seems that I can't do anything about it saying the app being damaged, unless I pay for an Apple Developer account to sign the app.

Not sure if this is the same issue, but with FutureRestore GUI, I just had CI do `codesign --remove-signature FutureRestore\ GUI.app` after jpackage which solved it from appearing as damaged. Then right click > Open worked fine.

**Reply by airsquared** [reply]
Thanks for the solution, it does seem to be working now.

**Comment by PencilNotPen** [comment]
The new saved devices menu looks good, thanks! Importing my presets from 2.5.5 didn't seem to bring Apnonces for devices which had them specified. Imported presets with specified Apnonces have the Specify Apnonce box ticked, but show no Apnonce in the box below. Not sure if the interface is not displaying it when it is imported successfully, or if it really is not imported with the rest of the preset information. Not the end of the world if it's just 2.5.5->3.0.0 as only have to re-add for max 10 devices, but hopefully they'll be preserved correctly through future updates, when users may have many device presets saved, and may not notice the Apnonces have gone, saving invalid blobs as a result. Cheers!

**Reply by EdgeXS** [reply]
importing from 10.15 without Apnonce works.  Seems to work.  Can restore occur without the Apnonce?

**Reply by PencilNotPen** [reply]
A12+ devices need to have specified an Apnonce for valid blobs. A11- don’t.

**Reply by airsquared** [reply]
I just fixed the bug with not importing the apnonce correctly, please try the latest beta.

**Comment by EdgeXS** [comment]
I'd add one request under the frequency its to specify a specific time for it to run.  Since i have a system that isn't always on which will be my blob machine and it helps to verify the process is working based on a specific time frame for traceability.

**Reply by airsquared** [reply]
In addition to the time you set it at, blobs are also automatically saved on every boot. If your computer is shut down when your interval passes blobs will automatically be saved on the next boot, this way you don't need to worry about if your computer is off.

**Reply by EdgeXS** [reply]
great thanks.  even better.

**Comment by bs286692** [comment]
For the Windows version:

1. The app no longer minimizes to the tray where it can run in the background.

2. Users should be able to change the sort order rather than defaulting to alphabetical sorting only.

3. Part of the app is cut off at the bottom and I I cannot resize the window.

4. Please put the version number in the file name for the Windows version as you have for the MacOS and Linux versions, e.g. blobsaver-windows-2.6.exe.

Thanks

**Reply by airsquared** [reply]
1. This is intentional, with the new update blobsaver does not run all the time in the background and is only running when it is actually saving blobs.
2. I'm planning on adding more control over sorting in a future version, maybe in v3.1.
3. Please try the latest beta to see if it fixes anything.
4. Done.

**Reply by bs286692** [reply]
1. imho, the way you had this implemented before was correct, intuitive, and consistent with how Microsoft and other Windows apps do things. It's what the notification area/tray is for: agents, helpers, monitors, and managers.

Ideally, users should be able to: run blobsaver in the foreground, continue running it in the background when minimized to the tray, and still be able to close the program completely from either state.

If users are forced to keep the app open in the foreground, it will take up valuable real estate on the task bar as well as the app switcher when this is not necessary (or even desirable).

Please consider reverting back to the old behavior, or at the very least provide an option so that users can minimize to the system tray and run in the background like before.

Thanks

**Reply by airsquared** [reply]
I believe you are misunderstanding; with the new update blobsaver does not need to be running at all to save blobs in the background. You can close the application entirely and it will still save blobs in the background.

When it is actually time to save blobs, it will automatically and quietly activate itself in the background and close itself after saving blobs without showing any UI. 

This way, while blobsaver is not doing anything in the background, no extra system resources will be used because the application will not even be running.

**Reply by bs286692** [reply]
So if the blobsaver daemon crashes, how will the user know?

Or what if the user accidentally forgets to start it? Or what if there's an unrelated problem with JRE and it doesn't launch properly when configured to run on Windows startup?

I'm not so sure I like the idea of not having visual confirmation that the daemon is running. There are so many things that could go wrong that aren't even the fault of blobsaver.

Think of it like an antivirus application: even though it launches a daily scan by itself on a schedule, there is still an icon in the system tray to let the user know the daemon is running in the background.

At the very least, please consider a configuration option to always show a icon in the notification area / minimize to the tray.

Thanks

**Reply by airsquared** [reply]
I think the difference here is that blobsaver does not have a daemon, there is effectively no such thing as blobsaver running in the background. In order to have a system tray icon, blobsaver needs to be running all the time which is not the case. When you press "start background", blobsaver is telling Windows to automatically launch blobsaver with some parameters at a set interval. You can even look at the [Windows Task Scheduler](https://www.windowscentral.com/how-create-automated-task-using-task-scheduler-windows-10) and see an entry for blobsaver to confirm this.

With the antivirus example: if you close the application, you are not actually quiting it as the application is still running, just in the background. This is not the case with the new blobsaver version, which actually quits itself.

That is true that the user would not be aware if it fails while in the background. I'm considering using either a log file or native notifications to address this.

**Reply by bs286692** [reply]
Hmm, then something on my end is not working as expected as no blobs are being saved. Maybe I am misunderstanding what is supposed to happen.

This is what I did:

1. I clicked on "Background settings", then selected all of my devices.

I got the "Testing device... Saving blobs... Success!" message dialogue for each device.

2. I moved those saved blobs to a safe place for backup.

3. I clicked "Change frequency" and set "Frequency to check" to "1 days".

4. I clicked on "Start Background", and I got the message:

> Background saving has been enabled. You can now close this application and it will continue saving blobs at the interval you set, and won't use any resources when it is not running.

But no new blobs were saved, even after five days. 

I also tried changing the frequency to something shorter, as well as leaving the application open, but no new blobs.

I checked the Windows Task Scheduler and don't see a new task created.

Manual saving is still working as expected.

Thanks

**Reply by airsquared** [reply]
This does look like a bug.

- Does "Force Check" work?
- Does automatically saving not even work when you have a lower frequency like 5 minutes?

**Reply by bs286692** [reply]
- No, "Force check" is disabled and greyed out.
- No, I have tried setting automatic saving to a really short interval, and it still does not save blobs.

Thanks

**Reply by bs286692** [reply]
> I think the difference here is that blobsaver does not have a daemon, there is effectively no such thing as blobsaver running in the background. In order to have a system tray icon, blobsaver needs to be running all the time which is not the case. When you press "start background", blobsaver is telling Windows to automatically launch blobsaver with some parameters at a set interval.

On a side note, if you are switching to Windows task scheduling, I think you should move away from the previous background/daemon terminology to be more consistent with what is actually happening, i.e. scheduling tasks.

That's what confused me the most as I was trying to troubleshoot.

So instead of:

> Background saving has been enabled. You can now close this application and it will continue saving blobs at the interval you set, and won't use any resources when it is not running.

You have something like:

> A recurring blobsaver task has been scheduled and will be executed according to the frequency you set.

Thanks

**Reply by airsquared** [reply]
I just updated the betas, please try them out to see if it fixes your problem.

**Reply by bs286692** [reply]
Hi,

I just tried the updated beta, and I got the following error:

> Error: Background saving is not enabled. Try again with the debug lo...

When I enabled the Window-->Debug Log, I got the following error:

> ERROR: The task XML is malformed.
> (1,40)::ERROR: unable to switch the encoding

I checked, and no new task was created on the filesystem or in the registry:

```
C:\Windows\System32\Tasks
C:\Windows\Tasks

HKLM\Software\Microsoft\Windows NT\CurrentVersion\Schedule\Taskcache\Tasks
HKLM\Software\Microsoft\Windows NT\CurrentVersion\Schedule\Taskcache\Tree
```

I got the same results logged in as Administrator.

Thanks

**Reply by airsquared** [reply]
Just updated them again, please try this latest one which should fix it now. Seems that I made a small typo on that last one.

**Reply by bs286692** [reply]
I'm getting the following error now, and no new tasks created:

> ERROR: The task XML is malformed.
> (1,2)::ERROR: incorrect document syntax

Thanks

**Reply by airsquared** [reply]
I think I found the problem, please try the latest one again.

**Reply by bs286692** [reply]
It seems to be working! I tried both starting and stopping background.

**Reply by airsquared** [reply]
Thanks for helping me test!

**Reply by bs286692** [reply]
No problem at all. It's awesome to have background saving working.

Question: What happens if there's an error saving, e.g. disk full, network unreachable, etc.? Will the scheduled task trigger blobsaver.exe to alert the user?

**Reply by airsquared** [reply]
It currently doesn't alert the user, but if there is was problem on the last run it will show in the Windows Task Scheduler. There should also be logs stored [here](https://stackoverflow.com/questions/61538232/where-windows-10-the-task-scheduler-store-log-file), but I haven't checked to confirm this for sure.

In the specific case of the network being unreachable, it will postpone the task until a network connection is achieved. (at least according to my understanding of the options in Windows Task Scheduler)

I am considering adding support for system notifications to alert the user if something goes wrong in a future version of blobsaver.

**Comment by joostiphone** [comment]
It says:

Error

"0x1111111111111111" is not a valid apnonce

**Reply by joostiphone** [reply]
Hmm I guess, never mind. I've now read it with Blobsaver which provides a much longer number. In the past I always set the 0x111.. number, but I guess we can't do that anymore?

**Reply by UInt2048** [reply]
@joostiphone That 0x1111111111111111 is your generator, not your APNonce. Don't confuse the two.

**Reply by joostiphone** [reply]
That makes sense, I guess lol. Thanks for this.

**Comment by Absolucy** [comment]
When 3.0 is released, could you perhaps submit a cask to Homebrew? So it can be installed via `brew install --cask blobsaver`?

**Reply by airsquared** [reply]
Created a [PR](https://github.com/Homebrew/homebrew-cask/pull/106435).

**Reply by airsquared** [reply]
blobsaver v3.0.1 is now in Homebrew cask.

**Comment by CoocooFroggy** [comment]
You may want to specify an image for the .dmg. [Otherwise, it's just Duke](https://imgur.com/a/rUFmmd4).

You can do this with `jpackage [...] --icon /blobsaverMacIcons.icns`.

**Comment by buitim** [comment]
I'm digging the update so far! Would it be possible to save the generator as a part of a saved device? Or maybe that's just a bug on my end.

**Reply by airsquared** [reply]
Realized I forgot to add that, it should be in the latest beta now.

**Comment by oooooooooooooooo** [comment]
When trying to read apnonce from jailbroken device instead of manually inputting it the device was booted into recovery mode and was not booted out. It is currently stuck in recovery mode. 

Device is iPhone 12 mini on ios 14.3 using the Unc0ver jailbreak

UPDATE: Managed to exit out of recovery mode using Reiboot. Being more than a little dumb I tried again in a non-jailbroken state. Same issue but once the device rebooted to recovery mode after Blobsaver tried exiting out I just used Reiboot again on the side and everything worked perfectly.

**Reply by airsquared** [reply]
Did it give an error or just silently fail?

**Reply by oooooooooooooooo** [reply]
Silent, it just stalled until blobsaver just stopped trying

**Reply by airsquared** [reply]
I think I've found the problem, please try the latest beta.

**Comment by ghost** [comment]
What if I don't have systemd (gentoo OpenRC users rise up)?

**Reply by airsquared** [reply]
If you don't have systemd, you can also manually set it up using cron if you like. Once you have everything setup for background saving, instead of clicking "start background", set up a cron job that calls the blobsaver executable with the `--background-autosave` argument.
